### ISSUE TITLE
Image quality metrics

### ISSUE THREAD

**Issue by torsteingrindvik** [tracel-ai/burn#4312]
# Feature description

Add metrics for assessing image quality, including deep learning metrics.

- [x] L1/MAE loss, [torch](https://docs.pytorch.org/docs/stable/generated/torch.nn.L1Loss.html), [torch-image-metrics](https://github.com/mdipcit/torch-image-metrics?tab=readme-ov-file#-supported-metrics)
  - Closed via  https://github.com/tracel-ai/burn/pull/4341
- [x] L2/MSE loss, [torch](https://docs.pytorch.org/docs/stable/generated/torch.nn.MSELoss.html), [torch-image-metrics](https://github.com/mdipcit/torch-image-metrics?tab=readme-ov-file#-supported-metrics)
  - Closed via https://github.com/tracel-ai/burn/pull/4341 
- [x] PSNR, [Wikipedia](https://en.wikipedia.org/wiki/Peak_signal-to-noise_ratio), [scikit-image](https://scikit-image.org/docs/0.25.x/api/skimage.metrics.html#skimage.metrics.peak_signal_noise_ratio), [PyTorch-ignite](https://docs.pytorch.org/ignite/generated/ignite.metrics.PSNR.html#ignite.metrics.PSNR), [torch-image-metrics](https://github.com/mdipcit/torch-image-metrics?tab=readme-ov-file#-supported-metrics)
  - Closed via https://github.com/tracel-ai/burn/pull/4379
- [x] SSIM, [Wikipedia](https://en.wikipedia.org/wiki/Structural_similarity_index_measure), [scikit-image](https://scikit-image.org/docs/0.25.x/api/skimage.metrics.html#skimage.metrics.structural_similarity), [PyTorch-ignite](https://docs.pytorch.org/ignite/generated/ignite.metrics.SSIM.html#ignite.metrics.SSIM), [torch-image-metrics](https://github.com/mdipcit/torch-image-metrics?tab=readme-ov-file#-supported-metrics)
  - Closed via https://github.com/tracel-ai/burn/pull/4396
- [x] MS-SSIM,  [lightning.ai](https://lightning.ai/docs/torchmetrics/stable/image/multi_scale_structural_similarity.html), [pytorch-msssim](https://github.com/VainF/pytorch-msssim)
  - Closed via https://github.com/tracel-ai/burn/pull/4555
- [x] Smooth L1, [torch](https://docs.pytorch.org/docs/stable/generated/torch.nn.SmoothL1Loss.html)
  - Closed via https://github.com/tracel-ai/burn/pull/4547
- [x] Gram Matrix Loss, [torch tutorial](https://docs.pytorch.org/tutorials/advanced/neural_style_tutorial.html#style-loss), [piq](https://github.com/photosynthesis-team/piq/blob/9948a52fc09ac5f7fb3618ce64b7086f5c3109da/piq/perceptual.py#L307)
  - Closed via https://github.com/tracel-ai/burn/pull/4595 
- [x] FID, [lightning.ai](https://lightning.ai/docs/torchmetrics/stable/image/frechet_inception_distance.html), [pytorch-fid](https://github.com/mseitzer/pytorch-fid), [torch-image-metrics](https://github.com/mdipcit/torch-image-metrics/blob/main/src/torch_image_metrics/metrics/dataset.py#L18)
  - Closed via https://github.com/tracel-ai/burn/pull/4644 
- [x] LPIPS, [lightning.ai](https://lightning.ai/docs/torchmetrics/stable/image/learned_perceptual_image_patch_similarity.html#), [torch-image-metrics](https://github.com/mdipcit/torch-image-metrics?tab=readme-ov-file#-supported-metrics)
  - Closed via https://github.com/tracel-ai/burn/pull/4403
- [x] DISTS, [lightning.ai](https://lightning.ai/docs/torchmetrics/stable/image/dists.html)
  - Closed via https://github.com/tracel-ai/burn/pull/4574
- [ ] A-FINE, [IQA-PyTorch](https://github.com/chaofengc/IQA-PyTorch/pull/277)

# Feature motivation

When training visual AI models the loss function may include various image metrics.
Users would have to re-implement these image metrics unless readily available.

## Examples

Take for example the impressive recent [ml-sharp](https://apple.github.io/ml-sharp/) which uses at least PSNR, SSIM, LPIPS, DISTS.

<img width="766" height="223" alt="Image" src="https://github.com/user-attachments/assets/44a586cf-4d3d-4e3a-b1d2-13bef73ed33b" />

**Comment by crutcher** [comment]
We already have MseLoss:
- https://burn.dev/docs/burn/nn/loss/struct.MseLoss.html

We could write MaeLoss easily, but we might as well go for the generalized Mean(L(P) Norm Error)Loss; where MAE := L(1) and MSE := L(2).

There's already a flexible norm stack, general with specializations in the linalg package that I firmed up last year:
- https://burn.dev/docs/burn/tensor/linalg/index.html

And looking at that, I put some polish on that with a PR out now:
- https://github.com/tracel-ai/burn/pull/4318

**Comment by crutcher** [comment]
It's possible that the current MseLoss is not configurable on dims appropriately for images; and the PSNR MAX operation (and probably some of the others) requires info about geometry.

**Comment by torsteingrindvik** [comment]
Marked L1, L2 done via https://github.com/tracel-ai/burn/pull/4341, thanks @softmaximalist

**Comment by softmaximalist** [comment]
I'll work on the SSIM metric if no one else is working on it. Leaving this comment to avoid duplicate work.

**Comment by koreaygj** [comment]
I'll take on the `LPIPS` metric.

**Comment by kvthr** [comment]
If no one is working on the DISTS metric, I'll take on that one.

**Comment by koreaygj** [comment]
@kvthr Are you still working on this? If not, can I take this issue?

**Comment by torsteingrindvik** [comment]
Impressive how quick this work is being done!
Added some more evals/losses, @laggui could you see if there are any you think shouldn't be in Burn scope (and feel free to remove them if so)?

I'll shortly argue for them here:

- MS-SSIM: I've had good use of this in own work, the paper has >9k citations which is a good indicator of merit imo: https://www.cns.nyu.edu/pub/eero/wang03b.pdf
- Smooth L1: I'm interested in view synthesis, this paper: https://arxiv.org/abs/2602.06478v1 in _Table 5: Ablation Study of REPA Distillation_ shows smooth L1 works well
- Gram matrix: Used in Apple's ml-sharp https://arxiv.org/pdf/2512.10685 , quote: "Our results show that adding the Gram-matrix loss significantly improves results". It's not hard to calculate though but imo. still nice of Burn to offer
- FID: Another recent Apple paper: https://arxiv.org/pdf/2411.07765 uses FID a lot for evals

**Comment by laggui** [comment]
I don't see any issues with more metrics 👍 we could potentially carve them out into a standalone `burn-metrics` crate that is used by `burn-train`, but not required for now.

**Comment by softmaximalist** [comment]
I will work on Smooth L1.

**Comment by softmaximalist** [comment]
I'll work on MS-SSIM.

**Comment by kvthr** [comment]
@koreaygj  I’m not, you can work on this.

**Comment by softmaximalist** [comment]
I will work on Gram Matrix Loss.

**Comment by torsteingrindvik** [comment]
I'm proposing A-FINE, it performs well in some cases where others do not:

<img width="2520" height="895" alt="Image" src="https://github.com/user-attachments/assets/e67e74b2-1bbf-478a-8635-c4c7753f89ba" />

- [arxiv](https://arxiv.org/pdf/2503.11221)
- [github.io page](https://tianhewu.github.io/A-FINE-page.github.io/)
- [official repo](https://github.com/ChrisDud0257/AFINE/tree/master)

Apache 2.0.

**Comment by cong-or** [comment]
Hey! I'm just finishing up the RNNT loss PR (#4623). Would it be okay if I had a go at FID? 

If so, any guidance on how you'd want the InceptionV3 dependency handled would be really helpful.

**Comment by laggui** [comment]
I don't think anyone else is working on it, so it's all yours!

A couple of vision metrics were added recently which are based on a pre-trained features, take a look at the LPIPS, DISTS and Gram matrix loss implementations. They all import pytorch pre-trained weights to initialize the metrics.

Looks like this is a common pattern for some vision metrics. We might want to generalize the pattern to re-use some of that logic eventually.

**Comment by cong-or** [comment]
> I don't think anyone else is working on it, so it's all yours!
> 
> A couple of vision metrics were added recently which are based on a pre-trained features, take a look at the LPIPS, DISTS and Gram matrix loss implementations. They all import pytorch pre-trained weights to initialize the metrics.
> 
> Looks like this is a common pattern for some vision metrics. We might want to generalize the pattern to re-use some of that logic eventually.

thanks, working on it!

**Comment by Capataina** [comment]
Can I tackle A-FINE? seems like its proposed but not being worked on?

**Comment by laggui** [comment]
Yeah it should be up for graps!

**Comment by Capataina** [comment]
Okay, perfect

**Comment by Capataina** [comment]
Hey, sorry for the silence on this one. Went deeper into the spec than I expected and wanted to come back with actual questions.

I've read through the LPIPS (#4403), DISTS (#4574), FID (#4644) and Gram Matrix (#4595) PRs as templates, looked at `burn-store`, and read the PyIQA A-FINE implementation. A few things seem different from the other image-quality metrics and I wanted to check direction before starting:

- A-FINE is CLIP ViT-B/32 plus 5 small head modules (naturalness, fidelity, two logistic calibrators, and a learnable `k` adapter).
- CLIP has no existing implementation in burn as far as I can tell. burn-nn has the building blocks (`TransformerEncoder`, `MultiHeadAttention`, `LayerNorm`) but no ViT or CLIP itself.
- OpenAI's `ViT-B-32.pt` is a TorchScript archive so `PytorchStore` can't read it directly. But PyIQA's `afine.pth` on Hugging Face bundles a fine-tuned CLIP under a top-level `finetuned_clip` key alongside the head shards, so we can load everything from one file with `with_top_level_key("...")`. PyIQA does this too since that fine-tuned CLIP is what A-FINE was trained on.

Questions before I start writing:

**1. PR size.** This one's going to be pretty massive given we're porting CLIP + 5 heads + loader + tests. Do you want it as one PR under `crates/burn-train/src/metric/vision/afine/`, or split, e.g. CLIP ViT as its own PR into a reusable spot and A-FINE on top? I'd lean one PR to match the backbone-inlining pattern in LPIPS/DISTS/FID, but future CLIP-based metrics (CLIP-IQA etc.) would benefit from a shared backbone.

**2. Weight hosting.** The Hugging Face mirror for `afine.pth` works with burn's downloader (HF returns a 302 and reqwest follows it). But this would be burn's first HF URL, all existing metrics use `download.pytorch.org` or GitHub raw. The file's also fairly big, around 600 MB. Is HF fine, or would you prefer we arrange a tracel-ai release mirror?

**3. Attention weights.** CLIP's checkpoint uses PyTorch's native `nn.MultiheadAttention` format with Q/K/V fused as one `in_proj_weight`. burn-nn's `MultiHeadAttention` has them separate. I was planning a small private `ClipQkvAttention` inside `afine/` (single `Linear` plus `chunk(3)` at forward) to keep the weight map one-to-one and match the inlined-backbone pattern. Sound right?

**4. QuickGELU.** CLIP's weights need QuickGELU (`x * sigmoid(1.702 * x)`), not standard GELU. Add it inside `afine/` alongside the backbone, or promote to `burn-nn/src/activation/` since any future CLIP-based metric will want it?

**5. Testing.** Matching LPIPS/DISTS/FID: mostly property tests (identity, asymmetry, finite, non-zero-for-different) under the `vision` feature, plus one or two `#[ignore = "downloads pre-trained weights"]` tests that call `init_pretrained`. Was also thinking of one extra `#[ignore]`'d test comparing our output to a captured PyIQA scalar on a deterministic pair, as insurance against silent QuickGELU or channel-order bugs. No precedent for that though, so happy to drop.

Thanks!

**Comment by laggui** [comment]
> Questions before I start writing:
> 
> **1. PR size.** This one's going to be pretty massive given we're porting CLIP + 5 heads + loader + tests. Do you want it as one PR under `crates/burn-train/src/metric/vision/afine/`, or split, e.g. CLIP ViT as its own PR into a reusable spot and A-FINE on top? I'd lean one PR to match the backbone-inlining pattern in LPIPS/DISTS/FID, but future CLIP-based metrics (CLIP-IQA etc.) would benefit from a shared backbone.

Having a CLIP ViT implementation on its own could be useful, but it's not essential to this goal for the metric. So bringing it into a single PR is fine.

> **2. Weight hosting.** The Hugging Face mirror for `afine.pth` works with burn's downloader (HF returns a 302 and reqwest follows it). But this would be burn's first HF URL, all existing metrics use `download.pytorch.org` or GitHub raw. The file's also fairly big, around 600 MB. Is HF fine, or would you prefer we arrange a tracel-ai release mirror?

Yeah this is starting to be larger territory 😄 An official HF link should be fine, as long as it's not just from any user but potentially from an organization. We don't mirror the other sources when importing such weights.

> **3. Attention weights.** CLIP's checkpoint uses PyTorch's native `nn.MultiheadAttention` format with Q/K/V fused as one `in_proj_weight`. burn-nn's `MultiHeadAttention` has them separate. I was planning a small private `ClipQkvAttention` inside `afine/` (single `Linear` plus `chunk(3)` at forward) to keep the weight map one-to-one and match the inlined-backbone pattern. Sound right?

I don't see any issues with a custom attention head module to fit the CLIP architecture.
 
> **4. QuickGELU.** CLIP's weights need QuickGELU (`x * sigmoid(1.702 * x)`), not standard GELU. Add it inside `afine/` alongside the backbone, or promote to `burn-nn/src/activation/` since any future CLIP-based metric will want it?

For now we can keep it local to the CLIP implementation. We could move it out eventually into `burn-nn`.

> **5. Testing.** Matching LPIPS/DISTS/FID: mostly property tests (identity, asymmetry, finite, non-zero-for-different) under the `vision` feature, plus one or two `#[ignore = "downloads pre-trained weights"]` tests that call `init_pretrained`. Was also thinking of one extra `#[ignore]`'d test comparing our output to a captured PyIQA scalar on a deterministic pair, as insurance against silent QuickGELU or channel-order bugs. No precedent for that though, so happy to drop.

Sounds good!

**Comment by Capataina** [comment]
Perfect, I've already started the implementation and everything seems to align well. I'll change the vit implementation to inline as well

**Comment by antimora** [comment]
Everything on the list is done, A-FINE included (#4894). Thanks everyone. Closing.
